File: go/arrow/memory/util_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package memory

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestRoundToPowerOf2(t *testing.T) {
	tests := []struct {
		v, round int
		exp      int
	}{
		{60, 64, 64},
		{122, 64, 128},
		{16, 64, 64},
		{64, 64, 64},
		{13, 8, 16},
	}
	for _, test := range tests {
		t.Run(fmt.Sprintf("v%d_r%d", test.v, test.round), func(t *testing.T) {
			a := roundToPowerOf2(test.v, test.round)
			assert.Equal(t, test.exp, a)
		})
	}
}

func TestIsMultipleOfPowerOf2(t *testing.T) {
	tests := []struct {
		v, d int
		exp  bool
	}{
		{200, 256, false},
		{256, 256, true},
		{500, 256, false},
		{512, 256, true},
	}
	for _, test := range tests {
		t.Run(fmt.Sprintf("%d_%d_%t", test.v, test.d, test.exp), func(t *testing.T) {
			got := isMultipleOfPowerOf2(test.v, test.d)
			assert.Equal(t, test.exp, got)
		})
	}
}

File: cpp/src/arrow/compute/exec/util_test.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "arrow/compute/exec/hash_join.h"
#include "arrow/compute/exec/schema_util.h"
#include "arrow/testing/gtest_util.h"
#include "arrow/testing/matchers.h"

using testing::Eq;

namespace arrow {
namespace compute {

const char* kLeftPrefix = "left.";
const char* kRightPrefix = "right.";

TEST(FieldMap, Trivial) {
  HashJoinSchema schema_mgr;

  auto left = schema({field("i32", int32())});
  auto right = schema({field("i32", int32())});

  ASSERT_OK(schema_mgr.Init(JoinType::INNER, *left, {"i32"}, *right, {"i32"}, kLeftPrefix,
                            kRightPrefix));

  auto output = schema_mgr.MakeOutputSchema(kLeftPrefix, kRightPrefix);
  EXPECT_THAT(*output, Eq(Schema({
                           field("left.i32", int32()),
                           field("right.i32", int32()),
                       })));

  auto i =
      schema_mgr.proj_maps[0].map(HashJoinProjection::INPUT, HashJoinProjection::OUTPUT);
  EXPECT_EQ(i.get(0), 0);
}

TEST(FieldMap, TrivialDuplicates) {
  HashJoinSchema schema_mgr;

  auto left = schema({field("i32", int32())});
  auto right = schema({field("i32", int32())});

  ASSERT_OK(schema_mgr.Init(JoinType::INNER, *left, {"i32"}, *right, {"i32"}, "", ""));

  auto output = schema_mgr.MakeOutputSchema("", "");
  EXPECT_THAT(*output, Eq(Schema({
                           field("i32", int32()),
                           field("i32", int32()),
                       })));

  auto i =
      schema_mgr.proj_maps[0].map(HashJoinProjection::INPUT, HashJoinProjection::OUTPUT);
  EXPECT_EQ(i.get(0), 0);
}

TEST(FieldMap, SingleKeyField) {
  HashJoinSchema schema_mgr;

  auto left = schema({field("i32", int32()), field("str", utf8())});
  auto right = schema({field("f32", float32()), field("i32", int32())});

  ASSERT_OK(schema_mgr.Init(JoinType::INNER, *left, {"i32"}, *right, {"i32"}, kLeftPrefix,
                            kRightPrefix));

  EXPECT_EQ(schema_mgr.proj_maps[0].num_cols(HashJoinProjection::INPUT), 2);
  EXPECT_EQ(schema_mgr.proj_maps[1].num_cols(HashJoinProjection::INPUT), 2);
  EXPECT_EQ(schema_mgr.proj_maps[0].num_cols(HashJoinProjection::KEY), 1);
  EXPECT_EQ(schema_mgr.proj_maps[1].num_cols(HashJoinProjection::KEY), 1);
  EXPECT_EQ(schema_mgr.proj_maps[0].num_cols(HashJoinProjection::OUTPUT), 2);
  EXPECT_EQ(schema_mgr.proj_maps[1].num_cols(HashJoinProjection::OUTPUT), 2);

  auto output = schema_mgr.MakeOutputSchema(kLeftPrefix, kRightPrefix);
  EXPECT_THAT(*output, Eq(Schema({
                           field("left.i32", int32()),
                           field("left.str", utf8()),
                           field("right.f32", float32()),
                           field("right.i32", int32()),
                       })));

  auto i =
      schema_mgr.proj_maps[0].map(HashJoinProjection::INPUT, HashJoinProjection::OUTPUT);
  EXPECT_EQ(i.get(0), 0);
}

TEST(FieldMap, TwoKeyFields) {
  HashJoinSchema schema_mgr;

  auto left = schema({
      field("i32", int32()),
      field("str", utf8()),
      field("bool", boolean()),
  });
  auto right = schema({
      field("i32", int32()),
      field("str", utf8()),
      field("f32", float32()),
      field("f64", float64()),
  });

  ASSERT_OK(schema_mgr.Init(JoinType::INNER, *left, {"i32", "str"}, *right,
                            {"i32", "str"}, kLeftPrefix, kRightPrefix));

  auto output = schema_mgr.MakeOutputSchema(kLeftPrefix, kRightPrefix);
  EXPECT_THAT(*output, Eq(Schema({
                           field("left.i32", int32()),
                           field("left.str", utf8()),
                           field("left.bool", boolean()),

                           field("right.i32", int32()),
                           field("right.str", utf8()),
                           field("right.f32", float32()),
                           field("right.f64", float64()),
                       })));
}

}  // namespace compute
}  // namespace arrow
